app/models/topic.rb

```diff
@@ -655,16 +654,17 @@ def add_moderator_post(user, text, opts=nil)
                               skip_validations: true,
                               custom_fields: opts[:custom_fields])
 
-    if (new_post = creator.create) && new_post.present?
-      increment!(:moderator_posts_count) if new_post.persisted?
-      # If we are moving posts, we want to insert the moderator post where the previous posts were
-      # in the stream, not at the end.
-      new_post.update_attributes!(post_number: opts[:post_number], sort_order: opts[:post_number]) if opts[:post_number].present?
+    new_post = creator.create
 
-      # Grab any links that are present
-      TopicLink.extract_from(new_post)
-      QuotedPost.extract_from(new_post)
-    end
+    increment!(:moderator_posts_count) if new_post.persisted?
+
+    # If we are moving posts, we want to insert the moderator post where the previous posts were
+    # in the stream, not at the end.
+    new_post.update_columns(post_number: opts[:post_number], sort_order: opts[:post_number]) if opts[:post_number].present?
```

Comment:
Hei @SamSaffron. I had a closer look at this and try to address your concern in 3f99265.

I also rebased the PR and got the tests green again. https://github.com/discourse/discourse/commit/477eb0591e55a6d89d6b1928d6ddd8575a3ced73 added a couple of new tests that had uncovered some more problems with the old patch.

Let me know.